File: graphs/NumEnclaves.java
class Solution {
    
    boolean isInBoundary = false;
    
    public int numEnclaves(int[][] grid) {
        int ans = 0;
        
        boolean[][] visited = new boolean[grid.length][grid[0].length];
        
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(grid[i][j]==1){
                    int cur = dfs(grid, i, j, visited);
                    
                    // System.out.println(i +" " + j + " "+ cur + isInBoundary);
                    
                    if(!isInBoundary) ans+=cur;
                    isInBoundary=false;
                }
            }
        }
        return ans;
    }
        
    public int dfs(int[][] grid, int sr, int sc, boolean[][] visited){
        if(sr>=0 && sc >=0 && sr<grid.length && sc<grid[0].length && grid[sr][sc] == 1 && !visited[sr][sc]){
            if(sr==0 || sc ==0 || sr == grid.length-1 || sc==grid[0].length-1) isInBoundary = true;
            visited[sr][sc] = true;
            return 1 + dfs(grid,sr-1,sc,visited) + dfs(grid,sr,sc+1,visited) + dfs(grid,sr+1,sc,visited) + dfs(grid,sr,sc-1,visited);
        } else {
            return 0;
        }
    }
}

File: graphs/numEnclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:

        ans = 0
        N = len(grid)
        M = len(grid[0])

        def dfs(row, col):
            if(row<0 or col<0 or row>=N or col>=M): return True, 0 
            if(grid[row][col] == -1 or grid[row][col] == 0): return False, 0

            grid[row][col] = -1

            onEdge = False
            ans = 1

            for dx, dy in [[-1, 0], [0, 1], [1, 0], [0, -1]]:
                nx, ny = row + dx, col + dy
                a, b = dfs(nx, ny)
                onEdge = onEdge or a
                ans += b
            
            return onEdge, ans

        
        ans = 0
        for row_index in range(N):
            for col_index in range(M):
                if(grid[row_index][col_index] == 1):
                    onEdge, count = dfs(row_index, col_index)
                    if(not onEdge): ans += count

        return ans 
